lightning/src/ln/onion_utils.rs

```diff
@@ -384,15 +382,25 @@ pub(super) fn process_onion_failure<T: secp256k1::Signing, L: Deref>(secp_ctx: &
 						|| error_code == 21; // Special case error 21 as the Route object is bogus, TODO: Maybe fail the node if the CLTV was reasonable?
 
 						let mut network_update = None;
+						let mut failing_route_hop = None;
 
 						if error_code & NODE == NODE {
 							network_update = Some(NetworkUpdate::NodeFailure { node_id: route_hop.pubkey, is_permanent: error_code & PERM == PERM });
 						}
 						else if error_code & PERM == PERM {
-							network_update = if payment_failed { None } else { Some(NetworkUpdate::ChannelClosed {
-								short_channel_id: path[next_route_hop_ix - if next_route_hop_ix == path.len() { 1 } else { 0 }].short_channel_id,
-								is_permanent: true,
-							})};
+							network_update = if payment_failed { None } else {
+								let failing_route_hop = if is_from_final_node { route_hop } else { &path[route_hop_idx + 1] };
+								Some(NetworkUpdate::ChannelClosed {
+									short_channel_id: failing_route_hop.short_channel_id,
+									is_permanent: true,
+								})
+							};
+							failing_route_hop = if payment_failed { None } else {
+								Some(FailingRouteHop {
+									path: path.clone(),
+									index: if is_from_final_node { route_hop_idx } else { route_hop_idx + 1 } as u16,
```

Comment:
I think we should at least make a tiny bit of effort to decide if the error code seems to be blaming the inbound or outbound channel, instead of always saying outbound for not-final-node and inbound for final-node.

Reply:
I was primarily matching the logic used for the `NetworkUpdate` above. Would we want to do something similar there?

Reply:
Yea, we should maybe do something similar there, and also map some things that are currently channel failures to other types of failures. The code here is pretty old and admittedly a bit lazy. This comment is maybe on the wrong line, though - I guess the bigger question is later on in the non-PERM blocks. Maybe its ok to handle this stuff later and just land it as-is for now, as long as we get the API where we want.